[Linalg] Add *Conv2D* matchers

Abhishek-Varma opened the pull request:
-- This commit is the fourth in the series of adding matchers
   for linalg.*conv*/*pool*. Refer: https://github.com/llvm/llvm-project/pull/163724
-- In this commit all variants of Conv2D convolution ops have been
   added.
-- It also refactors the way these matchers work to make adding more matchers concise.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
I missed it in the previous review. I think we don't need `\` because this is the last line.

```suggestion
                                        strides);
```

hanhanW commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -791,32 +1010,604 @@ bool isaConvolutionOpOfType<linalg::DepthwiseConv2DNchwChwOp>(
+// #filterMap = affine_map<(N, W, C, w) -> (C, w)>
+// #outputMap = affine_map<(N, W, C, w) -> (N, C, W)>
+template <>
+bool isaConvolutionOpOfType<linalg::DepthwiseConv1DNcwCwOp>(
```
huh, github/gi is doing bad diff here. :(

I'll assume these are not changed and review new ops for now. I'll do a local diff for these methods before I click approval.

Abhishek-Varma replied:
Yes, you're correct. The diff isn't doing a good job. This PR only adds all *Conv2D* ops.

hanhanW commented on mlir/lib/Dialect/Linalg/Transforms/Specialize.cpp:
```diff
@@ -279,6 +279,17 @@ static FailureOr<LinalgOp> specializeLinalgConvolutions(RewriterBase &rewriter,
+  CONV_OP_SPECIALIZER(linalg::Conv2DNhwcHwcfQOp);
+  CONV_OP_SPECIALIZER(linalg::Conv2DNhwgcGfhwcQOp);
+  CONV_OP_SPECIALIZER(linalg::Conv2DNgchwGfchwQOp);
+  CONV_OP_SPECIALIZER(linalg::Conv2DNhwgcGfhwcOp);
```
Test for the op is missing. I don't find `conv_2d_nhwgc_gfhwc`. We do have `conv_2d_nhwgc_gfhwc_q` though, but that belongs to the above op.

Abhishek-Varma replied:
Thanks for catching! Added it now. I've even added a slight refactor of the Conv2 ops to match their order of appearance to be same in both `Specialize.cpp` and `Utils.cpp`.

hanhanW reviewed (commented):
I haven't reviewed detailed implementation yet, just pointing out some missing tests. Github/git is not doing good at diff this time, so it'll take me some time to review it.

(also, I'm still thinking if we can refactor something or not, so we don't need many changes.)

hanhanW commented:
> Hi @hanhanW !
> 
> That was such a nice suggestion. I've addressed your comments!
> 
> I was trying to also see if there's a way for us to simply just initialize with "(N, H, W, C, h, w)" string using the `ConvMatchBuilder` and then we simply just pass the string versions of the indexing maps directly - so this way the caller doesn't need to initialize so many dims locally. But it turned out to be a bit more involved so I just stuck with the current suggestion. :D
> 
> Can you please take a look now ?

I'll be out for a week. QQ: can we split it to two PRs? One for introducing `ConvMatchBuilder` and use it for existing patterns, and the other (which can be this one) is for 2D convs that we adding in this PR?

Abhishek-Varma committed:
[Linalg] Add *Conv2D* matchers

-- This commit is the fourth in the series of adding matchers
   for linalg.conv/pool. Refer: #163724
-- In this commit all variants of Conv2D convolution ops have been
   added.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

Abhishek-Varma commented:
I've rebased the PR to use `ConvMatchBuilder`.

banach-space commented on mlir/test/Dialect/Linalg/convolution/roundtrip-convolution.mlir:
```diff
@@ -55,6 +55,149 @@ func.func @conv_2d(%in : tensor<?x?xf32>, %filter : tensor<?x?xf32>, %out : tens
 
 // -----
 
+func.func @conv_2d_nhwc_hwcf(%input: tensor<?x?x?x?xf32>, %filter: tensor<?x?x?x?xf32>, %output: tensor<?x?x?x?xf32>) -> tensor<?x?x?x?xf32> {
```
[nit]
```suggestion
// -----------------------------
// Convolution ops - 2D
// -----------------------------

func.func @conv_2d_nhwc_hwcf(%input: tensor<?x?x?x?xf32>, %filter: tensor<?x?x?x?xf32>, %output: tensor<?x?x?x?xf32>) -> tensor<?x?x?x?xf32> {
```

Similar suggestion elsewhere.

banach-space reviewed (commented):
Thanks, I've left some comments inline.

Abhishek-Varma committed:
Review comment by Andrzej v1.0

Abhishek-Varma commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -240,27 +240,71 @@ bool isReductionIterator(utils::IteratorType iteratorType) {
+  Operation *subOp1 = mulOp->getOperand(0).getDefiningOp();
+  if (!isa_and_present<arith::SubIOp, arith::SubFOp>(subOp1))
+    return false;
+  Operation *subOp2 = mulOp->getOperand(1).getDefiningOp();
```
Done.

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -240,27 +240,91 @@ bool isReductionIterator(utils::IteratorType iteratorType) {
+///       %input - %input_scalar
+///          where, %input_scalar can have optional upcast operation.
+static bool bodyMatcherForConvolutionOps(Value yieldVal, Block *body,
+                                         bool zeroPointOffset = false) {
```
[nit] `zeroPointOffset` sounds like a variable that contains the offset. Perhaps `containsZeroPointOffset` or `withZeroPointOffset` or `quantized`? Basically, something that makes it clearer that there are only two options: `true` and `false`.

hanhanW replied:
I just learned that Abhishek is out for a while, so I go ahead to address the comment. Let's land it before it becomes a stale PR?

banach-space reviewed (approved):
The high-level structure looks good. I've not had a chance to scan the finer details, but IMO we can just rely on tests.

LGTM % nit, thanks!

hanhanW committed:
address comments

Signed-off-by: hanhanW <hanhan0912@gmail.com>
